crates/wasmtime/src/runtime/store.rs

```diff
@@ -1982,6 +1970,26 @@ at https://bytecodealliance.org/security.
         )
     }
 
+    #[inline]
+    #[cfg(feature = "component-model")]
+    pub(crate) fn component_resource_state_with_instance(
+        &mut self,
+        instance: crate::component::Instance,
+    ) -> (
+        &mut crate::runtime::vm::component::CallContexts,
+        &mut crate::runtime::vm::component::ResourceTable,
+        &mut crate::component::HostResourceData,
+        &mut crate::vm::component::ComponentInstance,
+    ) {
+        let instance = unsafe { &mut *self[instance.0].as_mut().unwrap().instance_ptr() };
```

Comment:
Mind leaving a TODO here to scope this `unsafe`? I think we can possibly do better in the future but for now this is fine